cem/support-typedef-jsdoc-utils.js

```diff
@@ -1,70 +1,141 @@
+    .map((type) => `/**@type {${type}}*/const foo=null;`)
+    .map((src) => ts.createSourceFile('src', src, ts.ScriptTarget.ES2015, true))
+    .map((ast) => ast.statements[0].jsDoc[0])
+    .flatMap((jsDocNode) => getTypesFromJsDocCommentNode(jsDocNode, ts))
+    .forEach((type) => {
+      types.add(type);
+    });
+
+  return Array.from(types);
+}
+
 export function getTypesFromConstructor (constructorNode, ts, retrievePrivate = false) {
   const constructorTypes = new Set();
 
-  constructorNode.body.statements.forEach((node) => {
+  constructorNode.body.statements
+    .flatMap((node) => {
+      if (!isVarInitWithDoc(node, ts, retrievePrivate)) {
+        return [];
+      }
+      // Retrieve only from the first JsDoc
+      return getTypesFromJsDocCommentNode(node.jsDoc[0], ts);
+    })
+    .forEach((type) => {
+      constructorTypes.add(type);
+    });
 
-    if (!isVarInitWithDoc(node, ts, retrievePrivate)) {
-      return;
-    }
+  return Array.from(constructorTypes);
+}
 
-    // Retrieve only the first @type found
-    const rawType = node.jsDoc[0].tags[0].typeExpression.type;
+function getTypesFromJsDocCommentNode (node, ts) {
+  if (node.kind !== ts.SyntaxKind.JSDocComment) {
+    return [];
+  }
 
-    const isUnionType = rawType.kind === ts.SyntaxKind.UnionType;
-    const isTupleType = rawType.kind === ts.SyntaxKind.TupleType;
-    let types;
+  // Retrieve only the first @type found
+  const rawType = node.tags[0].typeExpression.type;
 
-    if (isUnionType) {
-      types = rawType.types;
-    }
-    else if (isTupleType) {
-      types = rawType.elements;
-    }
-    else {
-      types = [rawType];
-    }
+  const isUnionType = rawType.kind === ts.SyntaxKind.UnionType;
+  const isTupleType = rawType.kind === ts.SyntaxKind.TupleType;
 
-    types.forEach((type) => {
-      const foundType = findCustomType(type, ts);
-      if (foundType != null) {
-        constructorTypes.add(foundType);
-      }
-    });
+  let types;
+  if (isUnionType) {
+    types = rawType.types;
+  }
+  else if (isTupleType) {
+    types = rawType.elements;
+  }
+  else {
+    types = [rawType];
+  }
 
+  const typesSet = new Set();
+  types.forEach((type) => {
+    const foundType = findCustomType(type, ts);
+    if (foundType != null) {
+      typesSet.add(foundType);
+    }
   });
 
-  return Array.from(constructorTypes);
-
+  return Array.from(typesSet);
 }
 
+const BUILTIN_TYPES = ['String', 'Boolean', 'Number', 'BigInt', 'Date', 'Map', 'Set', 'WeakMap', 'WeakSet', 'Object'];
 export function findCustomType (nodeType, ts) {
+  if (nodeType == null) {
+    return null;
+  }
+
+  const nodeKind = nodeType.kind;
+  const typeName = nodeType.typeName?.getText();
 
-  // Type[] (not primitive[])
-  const isArrayType = nodeType.kind === ts.SyntaxKind.ArrayType && nodeType?.elementType?.kind === ts.SyntaxKind.TypeReference;
-  if (isArrayType) {
-    return nodeType.elementType.typeName.getText();
+  // Type[] or primitive[]
```

Comment:
suggestion: I have to admit that it took me a while to understand why we would like to go through another loop of the function especially if we have a primitive type. Now I get it, it has been made to handle nested arrays. Could you change the comment to reflect that it also finds the type for nested arrays? Also, `or primitive[]` can be quite confusing imo as it will return `null` anyway.